controllers/resource_controller.go

```diff
@@ -0,0 +1,542 @@
+	client       client.Client
+	log          logr.Logger
+	versionCache pkgaddon.VersionCacheClient
+}
+
+func NewResourceController(mgr manager.Manager, stopChan <-chan struct{}, addonversioncache pkgaddon.VersionCacheClient) (controller.Controller, error) {
+	r := &resourceReconcile{
+		client:       mgr.GetClient(),
+		log:          ctrl.Log.WithName(resourceControllerName),
+		versionCache: addonversioncache,
+	}
+
+	c, err := controller.New(resourceControllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &appsv1.Deployment{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &v1.Service{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &v1.ServiceAccount{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &v1.Namespace{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &v1.ConfigMap{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &rbac_v1.ClusterRole{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &rbac_v1.ClusterRoleBinding{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &appsv1.DaemonSet{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &appsv1.ReplicaSet{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &appsv1.StatefulSet{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &batchv1.Job{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+
+	if err := c.Watch(&source.Kind{Type: &batch_v1beta1.CronJob{}}, r.enqueueRequestWithAddonLabel()); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func (r *resourceReconcile) enqueueRequestWithAddonLabel() handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(a client.Object) []reconcile.Request {
+		var reqs = make([]reconcile.Request, 0)
+		var labels = a.GetLabels()
+		if name, ok := labels["app.kubernetes.io/name"]; ok && strings.TrimSpace(name) != "" {
+			if ok, v := r.versionCache.HasVersionName(name); ok {
+				reqs = append(reqs, reconcile.Request{NamespacedName: types.NamespacedName{
+					Name:      v.Name,
+					Namespace: v.Namespace,
+				}})
+			}
+		}
+		return reqs
+	})
+}
+
+func (r *resourceReconcile) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.log.Info("reconciling", "request", req)
+
+	namespace := &v1.Namespace{}
+	err := r.client.Get(ctx, req.NamespacedName, namespace)
```

Comment:
Would there be a better way then trying to blindly get the resource into our typed object?

We should look into apiutil.
```
gvk, err = apiutil.GVKForObject(obj, c.Scheme())
```

Reply:
8b20a4ac7d110ee87fc94c9cf01f2f89e1450c94

Reply:
I am not sure how does the obj (runtime.Object) come from, but the controller-runtime client either get typed or unstructured object, which is targeted in the impl. I would keep this implementation as it is.